Feature/123 othercontainercontorollerandusecases

taniguchiiqqq opened the pull request:
## Overviews of implementation
add the following use case and controller

1. delete container
2. find container
3. update container name
4. Add food to a container
5. Delete food from a container

## Review points
I'm so sorry to make such a big PR. I should've divided it into two parts: about containers and food.
Please let me know if you find any problems with my implementation of mockContainerRepository.
I wanted to verify if the UpdateContainerNameUseCase changes data in the mock Repository, but I can't.

`UpdateContainerNameUseCase.test.ts`
```typescript
// TODO: add the mock container into the mock repo instead of using spy.
vi.spyOn(mockContainerRepository, "find").mockReturnValueOnce(
	Promise.resolve(container),
);

const result = await useCase.execute({
	containerId: containerId.id,
	name: "new name",
});
expect(result.ok).toBeTruthy();
```

I want to rewrite the above test like this. `UpdateUserNameUseCase.test.ts`
```typescript:
// add the mock user into the mock repo.
mockUserRepository.pushDummyData(user);

const result = await updateUserNameUseCase.execute({
	userId: USER_ID,
	targetUserId: USER_ID,
	name: UPDATED_NAME,
});

expect(result.ok).toBeTruthy();

// read the user from the mock repo.
const nameUpdatedUser = await mockUserRepository.find(userId);
expect(nameUpdatedUser!.name.name).toBe(UPDATED_NAME);
```

taniguchiiqqq committed:
Feature/104 create repositories interfaces (#105)

* make unit and quantity unnecessary (#103)

* feat: update some arg unnecessary

* chore: format

* refactor: shorten code

* feat: add Container Repository Interface

taniguchiiqqq committed:
Feature/106 create Container Controller, UseCase and related files. (#110)

* make unit and quantity unnecessary (#103)

* feat: update some arg unnecessary

* chore: format

* refactor: shorten code

* feat: add CreateContainerInterface

* feat: add CreateContainerUsecase

* feat: add test and mock

* feat: add container DTO

* feat: add createContainerController

* refactor: change ContainerPropsInterface

* fix: add foodDto

* fix: container dto

* fix: delete unnecessary lines, chores

* feat: add default static method to create new container

Co-authored-by: hirohata <hitohata@users.noreply.github.com>

* chore: format

* fix: ContainerDto

Co-authored-by: hirohata <43802663+hitohata@users.noreply.github.com>

---------

Co-authored-by: hirohata <hitohata@users.noreply.github.com>
Co-authored-by: hirohata <43802663+hitohata@users.noreply.github.com>

taniguchiiqqq committed:
feat: add delete container usecase

taniguchiiqqq committed:
feat: add delete container test

taniguchiiqqq committed:
refactor: modify interface prop name

taniguchiiqqq committed:
feat: add delete container controller

taniguchiiqqq committed:
feat: add find container usecase

taniguchiiqqq committed:
feat: add find container controller

taniguchiiqqq committed:
feat: add test for find container usecase

taniguchiiqqq committed:
feat: add update container name usecase

taniguchiiqqq committed:
feat: add test for update container usecase

taniguchiiqqq committed:
feat: add update container name contoroller

taniguchiiqqq committed:
feat: export IFoodProps

taniguchiiqqq committed:
feat: add Add food to container usecase

taniguchiiqqq committed:
fix: delete unnecessary foodId arg

taniguchiiqqq committed:
feat: add test for add food to container usecase

taniguchiiqqq committed:
feat: add controller for add food to container usecase

taniguchiiqqq committed:
fix: the response

taniguchiiqqq committed:
Merge branch 'feature/123-othercontainercontorollerandusecases' of https://github.com/genesis-tech-tribe/nishiki-backend into feature/123-othercontainercontorollerandusecases

taniguchiiqqq committed:
feat: add delete Food from container usecase

taniguchiiqqq committed:
feat: add test for delete food from container usecase

taniguchiiqqq committed:
chore: delete unnecessary console.log

hitohata commented on backend/main/src/Group/Controllers/CreateContainerController.ts:
This controller should get a GroupID.

https://genesis-tech-tribe.github.io/nishiki-documents/web-api/index.html#tag/container/paths/~1containers/post

hitohata commented on backend/main/src/Group/Controllers/UpdateContainerNameController.ts:
```diff
@@ -0,0 +1,28 @@
+import { Controller } from "src/Shared";
+import { UpdateContainerNameUseCase } from "src/Group/UseCases/UpdateContainerNameUseCase/UpdateContainerNameUseCase";
+import { IContainerDto } from "src/Group/Dtos/ContainerDto";
```
This module is not used.

hitohata commented on backend/main/src/Group/Domain/Entities/Container.ts:
```diff
@@ -23,11 +28,17 @@ export class Container extends AggregateRoot<string, IContainerProps> {
+		return Ok(new Container(id, props));
+	}
+
+	static default(
```
Please add a comment that explains when should this method be called.

hitohata commented on backend/main/src/Group/UseCases/AddFoodToContainerUseCase/AddFoodToContainerUseCase.ts:
```diff
@@ -0,0 +1,96 @@
+		this.containerRepository = containerRepository;
+	}
+
+	public async execute(
```
I think it needs to check the requesting user check.

taniguchiiqqq replied:
@hitohata 
Thank you for your comment! That's true. I don't have a concrete idea now, but I think that the authorization feature will be realized like this. Can you check if the idea seems to work?

To Get the User ID from HTTP header
- create Service to get UserId from CognitoAuthToken
- call the service, passing header of HTTP requests in the Router
- call the Controller, passing the user ID

To Verify User
- **add canEdit(userId) method to Group entity**
- **find Group related to the Container in this Usecase.**
- **call canEdit() in this Usecase.**

In this PR, I'll implement the feature shown in bold.

hitohata replied:
@taniguchiiqqq 
Thank you for planning of implementation.
I think it would work. Let's go with that policy for now.

I also understand implementing the bold features in this PR. It sounds good.

hitohata commented on backend/main/src/Group/UseCases/CreateContainerUseCase/CreateContainerUseCase.ts:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * Create a new container. You can call this use case without a container name, and then the new container's name will be the default name.
+ * @throws: When you try to create a new container that already existing, the error will throw an error.
```
I could not understand when this error would be thrown. Additionally, that error can be checked easily,  I think it's better to handle it explicitly somehow using the result type.

hitohata commented on backend/main/src/Group/UseCases/CreateContainerUseCase/ICreateContainerUseCase.ts:
```diff
@@ -0,0 +1,8 @@
+import { UseCaseError } from "src/Shared";
+import { ContainerDomainError } from "src/Group/Domain/Entities/Container";
+
+export interface ICreateContainerUseCase {
```
The container must be bound to a group. Therefore, this use case needs to get a Group ID.

hitohata commented on backend/main/test/Group/Domain/Entities/Food.test.ts:
```diff
@@ -1,5 +1,5 @@
+import { Food, FoodDomainError, FoodId } from "src/Group/Domain/Entities/Food";
 import {
 	Quantity,
 	QuantityError,
```
This module is not used.

hitohata commented on backend/main/test/Group/UseCase/AddFoodToContainerUseCase.test.ts:
```diff
@@ -0,0 +1,60 @@
+import { MockContainerRepository } from "../MockContainerRepository";
+import { Container, ContainerId } from "src/Group/Domain/Entities/Container";
+import { ContainerIsNotExisting } from "src/Group/UseCases/AddFoodToContainerUseCase/IAddFoodToContainerUseCase";
+import { Food, FoodId } from "src/Group/Domain/Entities/Food";
```
These modules are not used.

hitohata commented on backend/main/test/Group/UseCase/AddFoodToContainerUseCase.test.ts:
```diff
@@ -0,0 +1,60 @@
+		vi.clearAllMocks();
+	});
+
+	it("change exiting container'name", async () => {
```
Is this a typo? container's name?

taniguchiiqqq replied:
@hitohata 
It was supposed to be 'change exiting container', but I'll change it to more specific like 'add food to a exiting container'

hitohata commented on backend/main/test/Group/MockContainerRepository.ts:
```diff
@@ -0,0 +1,33 @@
+	async find(id: ContainerId): Promise<Container | null> {
+		return (
+			this.memoryContainers.find((container) => {
+				container.id.equal(id);
```
You must return true to return the value.

```ts
if (container.id.equal(id)) return true;
```
or

remove the {} braket.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

hitohata commented on backend/main/test/Group/UseCase/CreateContainerUseCase.test.ts:
```diff
@@ -0,0 +1,40 @@
+import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
+import { IContainerRepository } from "src/Group/Domain/IContainerRepository";
```
This module is not used.

hitohata commented on backend/main/test/Group/UseCase/DeleteFoodFromContainerUseCase.test.ts:
```diff
@@ -0,0 +1,75 @@
+import { ContainerIsNotExisting } from "src/Group/UseCases/DeleteFoodFromContainerUseCase/IDeleteFoodFromContainerUseCase";
+import { Food, FoodId } from "src/Group/Domain/Entities/Food";
+
+describe("find container use case", () => {
```
This must be 'delete a food from container use case'.

hitohata commented on backend/main/test/Group/UseCase/AddFoodToContainerUseCase.test.ts:
```diff
@@ -0,0 +1,60 @@
+import { Quantity } from "src/Group/Domain/ValueObjects/Quantity";
+import { Unit } from "src/Group/Domain/ValueObjects/Unit";
+
+describe("find container use case", () => {
```
This must be 'add a food to container use case'.

hitohata commented on backend/main/test/Group/UseCase/UpdateContainerNameUseCase.test.ts:
```diff
@@ -0,0 +1,44 @@
+import { Container, ContainerId } from "src/Group/Domain/Entities/Container";
+import { ContainerIsNotExisting } from "src/Group/UseCases/UpdateContainerNameUseCase/IUpdateContainerNameUseCase";
+
+describe("find container use case", () => {
```
This must be 'update container name use case'.

hitohata reviewed (changes requested):
> Please let me know if you find any problems with my implementation of mockContainerRepository.

Please check the comment in the mock container repository.

You also need to check the test in the delete a food from the container. Its test has failed.

taniguchiiqqq committed:
fix: add groupId for create container Interface

taniguchiiqqq committed:
fix: add typedoc for default method

taniguchiiqqq committed:
fix: remove unused component

taniguchiiqqq committed:
fix: add groupId to controller arg

taniguchiiqqq committed:
fix: container repo mock

taniguchiiqqq committed:
fix: the condition had a bug
the equal operator '=' couldn't compare ID properly

taniguchiiqqq committed:
fix: push data to mock instead of spyOn

taniguchiiqqq committed:
fix: push data to mock instead of spyOn in Delete test

taniguchiiqqq committed:
refactor: make find method one liner

taniguchiiqqq committed:
fix: remove unused component

taniguchiiqqq committed:
fix: add food container test case name

taniguchiiqqq committed:
fix: add groupId to the create container test

taniguchiiqqq committed:
feat: add canEdit method

taniguchiiqqq committed:
feat: add GroupRepository with findByContainerId

taniguchiiqqq committed:
fix: add user validation

taniguchiiqqq committed:
feat: add MockGroupRepository

taniguchiiqqq committed:
fix: update test to care about user auth

taniguchiiqqq committed:
fix: error message

taniguchiiqqq committed:
fix: add user validation to create container

taniguchiiqqq committed:
fix: remove unnecessary module

taniguchiiqqq committed:
fix: test case description

taniguchiiqqq committed:
fix: add missed error types

taniguchiiqqq committed:
feat: user validation for deleteContainerUsecase

taniguchiiqqq committed:
feat: user validation for deleteFoodUsecase

taniguchiiqqq committed:
feat: user validation for FindContainerUsecase

taniguchiiqqq committed:
feat: user validation for UpdateContainerNameUsecase

taniguchiiqqq committed:
fix: DeleteFoodUsecase test

taniguchiiqqq committed:
feat: update interface of controllers

taniguchiiqqq committed:
fix: delete containerId from Group

taniguchiiqqq commented:
@hitohata 
I'm so sorry for bothering you with the too-big PR. Could you review the code that I modified again? most of the changes related to this feature.

> **To Verify User**
> add canEdit(userId) method to Group entity
> find Group related to the Container in this Usecase.
> call canEdit() in this Usecase.

In order to realize it, I added the `findByContainerId` method to IGroupRepogitory. So if these changes are approved, I'm gonna add an access pattern and GSI to the DynamoDB definition.

hitohata commented on backend/main/src/Group/Controllers/DeleteContainerController.ts:
```diff
@@ -0,0 +1,31 @@
+			return this.badRequest(result.error.message);
+		}
+
+		return this.accepted();
```
This should be no content.
related to the following:
https://github.com/orgs/genesis-tech-tribe/discussions/83

hitohata commented on backend/main/src/Group/Controllers/AddFoodToContainerController.ts:
```diff
@@ -0,0 +1,29 @@
+			return this.badRequest(result.error.message);
+		}
+
+		return this.accepted();
```
This should be no content.
related to the following:
https://github.com/orgs/genesis-tech-tribe/discussions/83

hitohata commented on backend/main/src/Group/Controllers/CreateContainerController.ts:
```diff
@@ -0,0 +1,32 @@
+			return this.badRequest(result.error.message);
+		}
+
+		return this.ok(result.value);
```
This return status should be "created".
https://genesis-tech-tribe.github.io/nishiki-documents/web-api/index.html#tag/container/paths/~1containers/post

hitohata replied:
This is a question.
I think this controller is going to be connected to the web API router, will you convert a container DTO to ID in the router? The return value is container ID only.

https://genesis-tech-tribe.github.io/nishiki-documents/web-api/index.html#tag/container/paths/~1containers/post

taniguchiiqqq replied:
@hitohata 
Thank you for your question. I would format data in the router because I thought it could be the presentation layer's responsibility. But now I feel I can also write it in this Controller. What do you say about it? I think either is fine but we need to decide it and make consistency.

hitohata replied:
@taniguchiiqqq 
Thank you for getting back to me.
I've thought twice. Then, you are right. The changing format is the responsibility of the presentation layer. And, this Controller can be called other domains in the same code base as a repository. So, your implementation is appropriate.

taniguchiiqqq replied:
@hitohata 
>  this Controller can be called other domains in the same code base as a repository
Thank you! It totally makes sense. I'll stick to this implementation.

hitohata commented on backend/main/src/Group/Controllers/UpdateContainerNameController.ts:
```diff
@@ -0,0 +1,28 @@
+			return this.badRequest(result.error.message);
+		}
+
+		return this.accepted();
```
This should be no content.
related to the following:
https://github.com/orgs/genesis-tech-tribe/discussions/83

hitohata commented on backend/main/src/Group/Domain/Entities/Container.ts:
```diff
@@ -77,7 +94,9 @@ export class Container extends AggregateRoot<string, IContainerProps> {
 	public removeFood(foodId: FoodId): Result<Container, ContainerDomainError> {
-		const foods = this.props.foods.filter((f) => f.id !== foodId);
+		const foods = this.props.foods.filter((f) => {
+			if (!f.id.equal(foodId)) return true;
```
This is a comment with self-reflection, you don't need to fix it. This value inside the iterator looks ambiguous. It should be 'food' or something like that.

hitohata commented on backend/main/src/Group/Domain/IGroupRepository.ts:
```diff
@@ -0,0 +1,10 @@
+
+export interface IGroupRepository {
+	find(id: GroupId): Promise<Group | null>;
+	findByContainerId(containerId: ContainerId): Promise<Group | null>;
```
This is a tip. You can use the overload.

taniguchiiqqq replied:
OMG, what a cool solution. I should've noticed this chance. I updated it. thank you so much

hitohata commented on backend/main/src/Group/Controllers/AddFoodToContainerController.ts:
```diff
@@ -0,0 +1,29 @@
+		const result = await this.useCase.execute({ ...input });
+
+		if (!result.ok) {
+			return this.badRequest(result.error.message);
```
This use case has the possibility of returning the `UserIsNotAuthorized` error. In that case, do you think it needs to return to `forbidden`? At least, I think so.

taniguchiiqqq replied:
Thank you! I overlooked it :pray: That's true.

hitohata commented on backend/main/src/Group/Controllers/CreateContainerController.ts:
```diff
@@ -0,0 +1,32 @@
+		const result = await this.useCase.execute({ userId, name, groupId });
+
+		if (!result.ok) {
+			return this.badRequest(result.error.message);
```
This use case has the possibility of returning the UserIsNotAuthorized error. In that case, do you think it needs to return to forbidden? At least, I think so.

hitohata commented on backend/main/src/Group/Controllers/DeleteContainerController.ts:
```diff
@@ -0,0 +1,31 @@
+
+		const result = await this.useCase.execute({ userId, containerId });
+
+		if (!result.ok) {
```
This use case has the possibility of returning the UserIsNotAuthorized error. In that case, do you think it needs to return to forbidden? At least, I think so.

hitohata commented on backend/main/src/Group/UseCases/DeleteContainerUseCase/DeleteContainerUseCase.ts:
```diff
@@ -0,0 +1,80 @@
+
+		await this.containerRepository.delete(containerId);
+
+		const newGroupOrError = group.removeContainerId(containerId);
```
Sorry, I couldn't point it out before. I feel this method name is appropriate if it is `removeContainer` instead of `removeContainerId`. It is my preference though.

taniguchiiqqq replied:
Thank you for mentioning it! I've updated it

hitohata commented on backend/main/src/Group/Controllers/FindContainerController.ts:
```diff
@@ -0,0 +1,31 @@
+		const result = await this.useCase.execute({ userId, containerId });
+
+		if (!result.ok) {
+			return this.badRequest(result.error.message);
```
This use case has the possibility of returning the UserIsNotAuthorized error. In that case, do you think it needs to return to forbidden? At least, I think so.

hitohata commented on backend/main/src/Group/UseCases/FindContainerUseCase/FindContainerUseCase.ts:
```diff
@@ -0,0 +1,72 @@
+		// check the user is the member of the group
+		const group = await this.groupRepository.findByContainerId(containerId);
+		if (!group) {
+			return Err(
```
Shouldn't this case return null?

taniguchiiqqq replied:
Oh, It makes sense. Thank you!

hitohata commented on backend/main/src/Group/Controllers/UpdateContainerNameController.ts:
```diff
@@ -0,0 +1,28 @@
+		const result = await this.useCase.execute({ userId, containerId, name });
+
+		if (!result.ok) {
+			return this.badRequest(result.error.message);
```
This use case has the possibility of returning the UserIsNotAuthorized error. In that case, do you think it needs to return to forbidden? At least, I think so.

hitohata commented on backend/main/test/Group/UseCase/AddFoodToContainerUseCase.test.ts:
```diff
@@ -0,0 +1,107 @@
+
+	it("User is not authorized", async () => {
+		const USER_ID_2 = UserId.generate().id;
+		const userId_2 = UserId.create(USER_ID_2).unwrap();
```
Why do you create a user ID again?
This USER_ID_2 does not work out of this scope, so you can use camelCase. In JS, the variable name should be camelCase, do not mix camelCase and snake_case.

taniguchiiqqq replied:
Oh my gosh, it's totally pointless. I also made the userId from USER_ID, so I will fix them.

hitohata commented on backend/main/test/Group/UseCase/CreateContainerUseCase.test.ts:
```diff
@@ -0,0 +1,89 @@
+	});
+	it("create container by invalid user", async () => {
+		const USER_ID_2 = UserId.generate().id;
+		const userId_2 = UserId.create(USER_ID_2).unwrap();
```
Why do you create a user ID again?
This USER_ID_2 does not work out of this scope, so you can use camelCase. In JS, the variable name should be camelCase, do not mix camelCase and snake_case.

hitohata commented on backend/main/test/Group/UseCase/DeleteContainerUseCase.test.ts:
```diff
@@ -0,0 +1,67 @@
+	});
+	it("User is not authorized", async () => {
+		const USER_ID_2 = UserId.generate().id;
+		const userId_2 = UserId.create(USER_ID_2).unwrap();
```
Why do you create a user ID again?
This USER_ID_2 does not work out of this scope, so you can use camelCase. In JS, the variable name should be camelCase, do not mix camelCase and snake_case.

hitohata commented on backend/main/test/Group/UseCase/DeleteFoodFromContainerUseCase.test.ts:
```diff
@@ -0,0 +1,123 @@
+	});
+	it("User is not authorized", async () => {
+		const USER_ID_2 = UserId.generate().id;
+		const userId_2 = UserId.create(USER_ID_2).unwrap();
```
Why do you create a user ID again?
This USER_ID_2 does not work out of this scope, so you can use camelCase. In JS, the variable name should be camelCase, do not mix camelCase and snake_case.

hitohata commented on backend/main/test/Group/UseCase/FindContainerUseCase.test.ts:
```diff
@@ -0,0 +1,106 @@
+		}).unwrap();
+
+		const USER_ID_2 = UserId.generate().id;
+		const userId_2 = UserId.create(USER_ID_2).unwrap();
```
Why do you create a user ID again?
This USER_ID_2 does not work out of this scope, so you can use camelCase. In JS, the variable name should be camelCase, do not mix camelCase and snake_case.

hitohata commented on backend/main/test/Group/UseCase/FindContainerUseCase.test.ts:
Your implementation should check the behavior of the group is not found case.

hitohata commented on backend/main/test/Group/UseCase/CreateContainerUseCase.test.ts:
You need to check the behavior of the group does not exist case.

hitohata commented on backend/main/test/Group/UseCase/AddFoodToContainerUseCase.test.ts:
You need to check the behavior of the group does not exist case.

hitohata commented on backend/main/test/Group/UseCase/UpdateContainerNameUseCase.test.ts:
```diff
@@ -0,0 +1,102 @@
+		}).unwrap();
+
+		const USER_ID_2 = UserId.generate().id;
+		const userId_2 = UserId.create(USER_ID_2).unwrap();
```
Why do you create a user ID again?
This USER_ID_2 does not work out of this scope, so you can use camelCase. In JS, the variable name should be camelCase, do not mix camelCase and snake_case.

hitohata reviewed (changes requested):
Please check the documentation and update it if it is needed.
This system doesn't work only with the backend code. We have to communicate with other teams asynchronously. To do so, the documentation is one of the tools. We have to provide correct information to other teams and we have to implement to accomplish it.

taniguchiiqqq committed:
Merge remote-tracking branch 'origin/develop' into feature/123-othercontainercontorollerandusecases

taniguchiiqqq committed:
fix: response status

taniguchiiqqq committed:
fix: add forbidden error

taniguchiiqqq committed:
fix: return null when the group not found

taniguchiiqqq committed:
chore: change user not authorized message

taniguchiiqqq committed:
fix: remove pointless UserId.create

taniguchiiqqq committed:
fix: add Group not found test cases

taniguchiiqqq committed:
refactor: change removeContainerId method name to

taniguchiiqqq committed:
refactor: using overload instead of findByContainerId

taniguchiiqqq committed:
fix: findByContainerId to find

hitohata reviewed (approved):
Think you for your implementation!

taniguchiiqqq commented:
Thank you for your careful reviews! I really appreciate it. 🙇 🙌
